File: 0733-flood-fill/0733-flood-fill.js
/**
 * @param {number[][]} image
 * @param {number} sr
 * @param {number} sc
 * @param {number} color
 * @return {number[][]}
 */

const dfs = (image, r, c, color, newColor) => {
    if (image[r][c] === color) {
        image[r][c] = newColor;
        if (r >= 1) dfs(image, r - 1, c, color, newColor);
        if (c >= 1) dfs(image, r, c - 1, color, newColor);
        if (r + 1 < image.length) dfs(image, r + 1, c, color, newColor);
        if (c + 1 < image[0].length) dfs(image, r, c + 1, color, newColor);
    }
}

const floodFill = (image, sr, sc, color) => {
    const oldColor = image[sr][sc];
    if (oldColor !== color) dfs(image, sr, sc, oldColor, color);
    return image;
};

File: 0733-flood-fill/0733-flood-fill.rb
# @param {Integer[][]} image
# @param {Integer} sr
# @param {Integer} sc
# @param {Integer} color
# @return {Integer[][]}
def flood_fill(image, sr, sc, color)
    originalColor = image[sr][sc]
    def dfs(r, c, originalColor, image, color)
        if r < 0 || c < 0 || r >= image.length || c >= image[0].length
            return
        end
        currColor = image[r][c]
        if currColor != originalColor || currColor == color
            return
        end
        image[r][c] = color
        dfs(r + 1, c, originalColor, image, color)
        dfs(r - 1, c, originalColor, image, color)
        dfs(r, c + 1, originalColor, image, color)
        dfs(r, c - 1, originalColor, image, color)
    end
    dfs(sr, sc, originalColor, image, color)
    return image
end